### PR TITLE
feat(explore): support EDS attributes

### PR DESCRIPTION
**Pull request by skjindal93**
## Description
Supports EDS attributes for explore for entity types

### Assumptions
- Presence of a group by in the explore query. If not, we can directly use the `entities` query for filters, selections, order bys
- Filters (_apart from inherent time range filter_), group bys, selections, order bys should be from a single source, either from QS or EDS

### Future Use Cases
- Can be extended in future to support filters on QS easily
- Can be extended in future to support selections, order bys on multiple sources, by registering different entity handlers

### Testing
Deployed the query pipeline locally, and tested the UI screens

### Checklist:
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/visitor/ExecutionVisitor.java [hunk 1]
```diff
@@ -46,16 +44,17 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(ExecutionVisitor.class);
 
   private final EntityQueryHandlerRegistry queryHandlerRegistry;
+  private final EntitiesRequest entitiesRequest;
   private final ExecutionContext executionContext;
-  private static final Logger LOG = LoggerFactory.getLogger(ExecutionVisitor.class);
 
   public ExecutionVisitor(
-      ExecutionContext executionContext, EntityQueryHandlerRegistry queryHandlerRegistry) {
+      EntitiesRequest entitiesRequest,
```
Comment (avinashkolluru) [inline gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/visitor/ExecutionVisitor.java:54]: Didnt understand the motivation for all these changes. Why did you have to pass the EntitiesRequest explicitly here where it is there in ExecutionContext?

File: gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionContext.java [hunk 2]
```diff
@@ -75,30 +82,51 @@
   private final Set<String> pendingMetricAggregationSourcesForOrderBy = new HashSet<>();
   private boolean sortAndPaginationNodeAdded = false;
 
-  // map of filter, selections (attribute, metrics, aggregations), order by attributes to source map
+  // map of filter, selections (attribute, metrics, aggregations), group bys, order by attributes to
+  // source map
   private final Map<String, Set<String>> allAttributesToSourcesMap = new HashMap<>();
 
-  private ExecutionContext(
+  public ExecutionContext(
```
Comment (avinashkolluru) [inline gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionContext.java:89]: Not clean. Looks like we changed this for the purpose of reusing some part of the code

File: gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionTreeUtils.java [hunk 3]
```diff
@@ -13,94 +12,51 @@
...
    * Returns a non-empty optional if all the attributes in the selection(attributes and
    * aggregations), time aggregations, filter and order by can be read from the same source.
    *
-   * @param executionContext
+   * @param expressionContext
    * @return
    */
-  static Optional<String> getSingleSourceForAllAttributes(ExecutionContext executionContext) {
-    Optional<String> singleSourceFromKeySets = getSingleSourceFromKeySets(executionContext);
+  public static Optional<String> getSingleSourceForAllAttributes(
```
Comment (avinashkolluru) [inline gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionTreeUtils.java:25]: Dont remember why we added these methods as utils. Can't these be defined within ExpressionContext?

File: gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionContext.java (hunk 2 above)
Comment (skjindal93) [reply gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionContext.java]: Actually no. I just got rid of the other static `from` method required to instantiate execution context

File: gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionTreeUtils.java (hunk 3 above)
Comment (skjindal93) [reply gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionTreeUtils.java]: Can move them inside `ExpressionContext` too

**Review by avinashkolluru — APPROVED** [review]

File: gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/EntityExecutionContext.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+  }
+
+  public List<Expression> getEntityIdExpressions() {
+    List<String> entityIdAttributeNames =
+        AttributeMetadataUtil.getIdAttributeIds(
+            attributeMetadataProvider,
+            entityIdColumnsConfigs,
+            this.entitiesRequestContext,
+            entitiesRequestContext.getEntityType());
+    return IntStream.range(0, entityIdAttributeNames.size())
```
Comment (rish691) [inline gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/EntityExecutionContext.java:137]: nit: why not just `entityIdAttributeNames.stream().map(...).collect(Collectors.toList())`

File: gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/EntityExecutionContext.java [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        sourceToSelectionExpressionMap.get(source).stream()
+            .filter(retainExpressionPredicate)
+            .collect(Collectors.toUnmodifiableList());
+
+    Map<String, List<Expression>> sourceToRetainedSelectionExpressionMap =
+        sourceToSelectionExpressionMap.entrySet().stream()
+            .map(
+                entry ->
+                    entry.getKey().equals(source) ? Map.entry(entry.getKey(), expressions) : entry)
+            .filter(entry -> !entry.getValue().isEmpty())
```
Comment (rish691) [inline gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/EntityExecutionContext.java:179]: nit: we can filter first

File: gateway-service-impl/src/main/java/org/hypertrace/gateway/service/common/ExpressionContext.java [hunk 6]
```diff
@@ -454,17 +338,256 @@ private void getSourceToFilterExpressionMap(
...
+   * ("API.id" -> ["EDS"], "API.discoveryState" -> ["QS"])
+   *
+   * The intersecting source set across all the attributes would be []
+   *
+   * </pre>
+   */
+  private static Set<String> getIntersectingSourceSets(
+      Map<String, Set<String>> attributeToSourcesMapFirst,
+      Map<String, Set<String>> attributeToSourcesMapSecond) {
+    if (attributeToSourcesMapFirst.isEmpty() && attributeToSourcesMapSecond.isEmpty()) {
```
Comment (rish691) [inline gateway-service-impl/src/main/java/org/hypertrace/gateway/service/common/ExpressionContext.java:542]: we want the intersection of values, can we do
```
Stream.concat(
    attributeToSourcesMapFirst.values().stream(), 
    attributeToSourcesMapSecond.values().stream())
.reduce(Sets::intersection)
.orElse(Collections.emptySet());
```

Comment (skjindal93) [reply gateway-service-impl/src/main/java/org/hypertrace/gateway/service/common/ExpressionContext.java:542]: An existing code, being refactored to a different class

**Review by rish691 — APPROVED** [review]

**Pull request merged**